Rewrite EndnoteXMLImporter as a StAX-Parser

DinjerChang opened the pull request:
Fixes #9538

<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->
This fixes https://github.com/JabRef/jabref/issues/9538 by rewriting the org.jabref.logic.importer.fileformat.EndnoteXMLImporter class to use a StAX-Parser, and removes all JAXB dependencies from the class. The corresponding xjc task is also removed from build.gradle.


<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

```[tasklist]
### Compulsory checks
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.
```

DinjerChang commented:
Hello, I have question about the expected format output of keywords. The expected keyword from *EndnoteXmlImporterTest_EmptyKeywordStyle.xml* has a semicolon and whitespace at the end: 
`keywords = {anxiety; craving; dependency; destress; health restoration; }`. However, from *EndnoteXmlImporterTestArticle.xml* the expected output of keywords field has no semicolon and whitespace at the end: `keywords = {Age Factors; Aged; Aged, 80 and over; Female; Great Britain/epidemiology; Humans; Male; Middle Aged; Models, Statistical; Neoplasms/*epidemiology; Risk Assessment; Risk Factors; Sex Characteristics}`. The difference cause the unit test to fail. 

My Implementation to realize keywords:
I store keywords in a List while parsing. After parsing finished, I use the built-in method `entry.putKeywords(keywords, preferences.bibEntryPreferences().getKeywordSeparator());` just like the original code and all of my keywords output (actual) will not have semicolon and whitespace at the end.

koppor commented:
(on the road, therefore brief reply). Good to have no end whitespace. You can change the test cases accordingly. JabRef has a class KeywordList (or similar) to handle keywords. Maybe that could be used here too? #reuse

Siedlerchr commented:
Please also make sure to import and setup JabRef's code style https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html

DinjerChang commented:
@koppor I'll look up the KeywordList (or similar), thank you.  As for the test cases, I'm not quite understand what "change the test cases accordingly" mean. As far as I known, the ImporterTestEngine replace the .xml source file with .bib file and create the expected output by BibtexParser. It would be great to have more explanation.

@Siedlerchr Sorry I forgot to run checkstyle before PR, will do it

Siedlerchr commented:
@DinjerChang  That sounds good:) Regarding the test case, you should modify the xml file then in the folder: 
https://github.com/JabRef/jabref/tree/main/src/test/resources/org/jabref/logic/importer/fileformat

koppor commented:
The change in CHANGELOG.md can be reverted. End users will see no change. In case more data from EndNote is read, that needs to be mentioned in the CHANGELOG.md. Otherwise, technical changes are not important for end users.

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/EndnoteXmlImporter.java:
```diff
@@ -174,156 +219,265 @@ private static EntryType convertRefNameToType(String refName) {
-                           .flatMap(url -> OptionalUtil.toStream(getUrlValue(url)))
-                           .findFirst();
+            if (isEndXMLEvent(reader) && reader.getName().getLocalPart().equals(startElement)) {
+//                System.out.println("keywords end");
```
You can use `LOGGER.debug` if this is important. Otherwise, remove the line.

koppor commented:
> @DinjerChang That sounds good:) Regarding the test case, you should modify the xml file then in the folder: https://github.com/JabRef/jabref/tree/main/src/test/resources/org/jabref/logic/importer/fileformat

I think, the XML is an EndNote file. We should not change the EndNote file. The expected BibTeX should be modified. See screenshot:

![image](https://github.com/JabRef/jabref/assets/1366654/8d5cfe66-a129-4a33-9cec-fc0c0e55a69e)

I think, only the line at https://github.com/JabRef/jabref/blob/2ee5c9cf3fbb095edf71d6d9044400188ebe0c0b/src/test/resources/org/jabref/logic/importer/fileformat/EndnoteXmlImporterTest_EmptyKeywordStyle.bib#L15 has be changed.

DinjerChang commented:
@koppor I have changed the .bib file accordingly, passed all the test case, removed redundant print statement and ran checkstyle to make sure it's aligned with the coding style. 
Just two question, so I can remove the line I  added in `CHANGLOG.md`? and do I have to create a new pull request after committing and pushing to my forked jabref repo? Or I could just simply commit and push, and then the pull request will automatically update my commit records here

Siedlerchr commented:
Yes, just remove the line and push the new commits. The PR gets automatically updated when you push your new commeits.

DinjerChang commented:
Do I have to fix the conflict listing above?

Siedlerchr commented:
Yes, you have to resolve the conflicts, otherwise the PR cannot be merged and tests won't run. It's probably just some formatter things which came from some other PRs recently. 
Best is to merge upstream/main into your branch and then resolve the conflicts.

Siedlerchr commented:
Looks good so far, there is only one additional code style checker failure: 
The good news is, simply running the task `rewriteRun` will fix it automatically :)

BUILD FAILED in 2m 9s
    org.jabref.config.rewrite.cleanup
        org.openrewrite.java.cleanup.EqualsAvoidsNull
Report available:
    /home/runner/work/jabref/jabref/build/reports/rewrite/rewrite.patch
Run 'gradle rewriteRun' to apply the recipes.

DinjerChang commented:
@Siedlerchr and @koppor  Thank you guys so much for the guidance! Just two more question:
1. So in addition to run checkstyle, we also have to run `gradle rewriteRun` everytime before making a pull request for the coding style of Jabref ?
2. Another subject: Our uni project require us to make pull request of test cases, I'm wondering if it's ok for us to make that. And if yes, is there any test cases need to be written for Jabref and under which feature/functionality do you recommend since I feel like the EndnoteXMLImpoter has a thorough test engine and need no test cases anymore and most of the issues are not test-cases-related.

Siedlerchr commented:
@DinjerChang  No, check style is enough. The openRewrite is a new thing. The CI will tell you (if you click on the failed action on details) if you need to run it.

For issues in general, we try to group and order them by e,g, difficulty or complexity: You can always take a loook here 
https://github.com/orgs/JabRef/projects 
Maybe that helps you find something. 

I have asked the other maintainers if they have something in mind or can come up with some that are easily testable. For tests, classes that reside in logic or model can be tested and already contain tests.

DinjerChang commented:
@Siedlerchr Thanks! I'll look aorund the link and the classes you mentioned.

Siedlerchr commented:
@DinjerChang  A good idea would be to craft some more tests for the citation key generator, especially the authors formatter see @koppor s work for how to do this  https://github.com/JabRef/jabref/pull/9799

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/EndnoteXmlImporter.java:
```diff
@@ -339,3 +502,8 @@ public List<BibEntry> parseEntries(InputStream inputStream) {
         return Collections.emptyList();
     }
 }
+
```
These are too many empty llines at the end. OK for me now to keep things going.

DinjerChang replied:
Ok, will pay attention to this next time as well.

koppor reviewed (approved):
For me, this improves the state-of-the art. Thus, +1 for merging

koppor commented:
Regarding tests. DId you know that IntelliJ supports running with code coverage?

![grafik](https://github.com/JabRef/jabref/assets/1366654/d2f844e8-4976-4bd1-9c3d-868fd25966bf)

I found out that the linked files handling has not been tested:

![grafik](https://github.com/JabRef/jabref/assets/1366654/61706e68-71fe-41e1-9339-d01a9cd45154)

I think, this is a hard one, because you don't have EndNote at hand and cannot check whether linked files handling properly works.

DinjerChang commented:
@koppor Thanks for the heads up regarding test. Yeah, we've been using the coverage technique to look around! We're working on some other part of the test cases
